### ISSUE TITLE
actions/upload-artifact & actions/download-artifact v4 support

### ISSUE THREAD

**Issue by Gedochao** [dorny/test-reporter#363]
## Describe
The popular https://github.com/actions/upload-artifact & https://github.com/actions/download-artifact recently got a major update to `v4`, which is incompatible with `v3` actions. 
The `v4` release offers a considerable boost in the efficiency of artifacts' uploads & downloads, so it's generally a much desired bump.
By the look of things, this breaks the flow for using the `test-reporter` with public repositories, as `test-reporter` uses those with `v3`.

## Example of a bump which seems to break the test reports flow (Scala CLI repo)
I did not do in-depth research on this topic, but it seems that bumping to `v4` in the Scala CLI repo breaks the test reports:
- https://github.com/VirtusLab/scala-cli/pull/2701

We use the public repo setup like this:
https://github.com/VirtusLab/scala-cli/blob/3dcfb3357d9c35bd71be27548e7e4e82fc6ef528/.github/workflows/test-report.yml
So we generate test reports with Mill, use our custom script to change them into the JUnit XML format and then upload the artifacts for the test report workflow to pick it up, nothing special. Should be seen as the standard `java-junit` flow for a public repo.

## Proposed solution
Bump `test-reporter` actions to `v4` & probably release a `v2` version of the `test-reporter` action, to prevent breaking old workflows and let them bump their stuff in peace.

## Alternatives considered
Stay at `v3`, I guess?

**Comment by Gedochao** [comment]
Extra context: this is sneaky, as it's very easy to bump to `v4` without reading the release notes, thus only realising the test reports are broken when inspecting them.
Users may also be confused when setting up the action for the first time and realising it doesn't work, not knowing why (so as a temporary measure, maybe it'd be useful if the `test-reporter@v1` docs mentioned the incompatibility 🤔

**Comment by ScribbleTAS** [comment]
Just updated to v4 and indeed, test-reporter is causing issues...
```
Error: HTTPError: Response code 400 (Authentication information is not given in the correct format. Check the value of Authorization header.)
```
Will downgrade again, but I'm looking forward to v2...

And I agree, having a
> [!WARNING]
> actions/upload-artifact & actions/download-artifact  **version 4** is currently not supported, downgrade to **v3**

or something like that, would really help clear up any confusion

**Comment by lauxjpn** [comment]
As a workaround, we manually download and extract our test results containing artifact in a previous step (that we uploaded before using our CI workflow and `actions/upload-artifact@v4`) using `actions/download-artifact@v4` or an `actions/github-script@v7` [script](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#using-data-from-the-triggering-workflow), and then use `dorny/test-reporter@v1` without the `artifact` option. That way, `path` just represents the local file system instead the one contained in an artifact.

https://github.com/dorny/test-reporter/blob/eaa763f6ffc21c7a37837f56cd5f9737f27fc6c8/src/main.ts#L82-L92

**Comment by jozefizso** [comment]
The #438 should fix the problem reported by @ScribbleTAS. The original issue will need another fix I think.

**Comment by JojOatXGME** [comment]
> The https://github.com/dorny/test-reporter/pull/438 should fix the problem reported by @ScribbleTAS. The original issue will need another fix I think.

@jozefizso and @Gedochao, what is the original issue? If there is another issue here, it is not clear to me what it is. I think the action should work with Artifacts v4.

Background: As far as I know, this action doesn't use `@actions/artifacts`, but instead communicates with the [GitHub REST API](https://docs.github.com/en/rest/actions/artifacts?apiVersion=2022-11-28#download-an-artifact). The backward incompatible changes which have been announced by GitHub for Artifacts v4 should not affect the GitHub REST API, but only some Runner-specific API to access Artifacts while the workflow was still in progress. Note that before Artifacts v4, Artifacts were not available over the GitHub REST API before the workflow has finished. Besides all of that, I suspect that GitHub made some internal architectural changes with Artifact v4 which caused the error reported by @ScribbleTAS. I am not aware of any other issue related to Artifacts v4.

**Comment by ScribbleTAS** [comment]
Hi @jozefizso and @JojOatXGME ,
thank you for working on this issue.

I have been testing the new version and it fixes the error, but I'm getting a new error:
```
Error: HttpError: Resource not accessible by integration
```

This error also shows up with upload v3

## Setup
I use the ["Recommended setup for public repositories"](https://github.com/dorny/test-reporter?tab=readme-ov-file#recommended-setup-for-public-repositories) from the readme, on a PR.

### build.yml

```yml
name: Build

on: [pull_request]

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3.5.0
    - name: Set up JDK 17 for x64
      uses: actions/setup-java@v3
      with:
        java-version: '17'
        distribution: 'temurin'
        architecture: x64
    - name: Setup Gradle
      uses: gradle/gradle-build-action@v2.4.0
      with:
        gradle-version: 8.6
    - name: Build TASmod with Gradle
      run: gradle build
    - name: Upload Test Report
      uses: actions/upload-artifact@v4
      if: always()
      with:
        name: TestResult
        path: build/test-results/test/*.xml
    - name: Upload TASmod-Dev-Build
      uses: actions/upload-artifact@v4
      with:
        name: TASmod-Dev-Build
        path: build/libs
```

> [!NOTE] 
> I know some of the other actions are outdated and I hope it's not the issue here...

### test_report.yml
```yml
name: 'Test Report'
on:
  workflow_run:
    workflows: ['Build']                     # runs after CI workflow
    types:
      - completed
jobs:
  report:
    runs-on: ubuntu-latest
    steps:
    - uses: dorny/test-reporter@v1.9.0
      with:
        artifact: TestResult
        name: TASmod Tests
        path: '*.xml'
        reporter: java-junit
```
## Tests
I ran a couple of tests on different versions now so here are my restults:
### Upload v3, TR v1.9.1
```
Error: HttpError: Resource not accessible by integration
```
Workflow run with debug: https://github.com/ScribbleTAS/TASmod/actions/runs/9016365464/job/24772939089
### Upload v4, TR v1.9.1
```
Error: HttpError: Resource not accessible by integration
```
Workflow run: https://github.com/ScribbleTAS/TASmod/actions/runs/9016031318/job/24771802152
### Upload v4, TR v1.9.0
```
HTTPError: Response code 400 (Authentication information is not given in the correct format. Check the value of Authorization header.)
```
Workflow run: https://github.com/ScribbleTAS/TASmod/actions/runs/9016105434/job/24772029996

**Comment by ScribbleTAS** [comment]
> I know some of the other actions are outdated and I hope it's not the issue here...
Just to be safe, I have updated all actions to their latest versions and v1.9.1 is still unusable for me...

**Comment by JojOatXGME** [comment]
@ScribbleTAS Since you have edited your comment and removed the block about your issue, I assume you have resolved it.

> I have been testing the new version and it fixes the error, but I'm getting a new error:
>
>     Error: HttpError: Resource not accessible by integration

For everyone else, the error did occur after the artifacts have already been downloaded and when the report was about to be created. While I don't know much about the case of @ScribbleTAS, this error usually means that your job doesn't have the `checks: write` permission. (Either because you configured [restricted permissions by default](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token) and forgot to specify the permission in the workflow file, or because the build was triggered from a fork and you did not follow the [recommended setup for public repos](https://github.com/dorny/test-reporter#recommended-setup-for-public-repositories).)

PS: Also note that `dorny/test-reporter@v1` still serves 1.9.0 which does not contain the fix. See #444. I just wanted to mention it as it might cause some confusion if someone wants to try out the new version.

**Comment by ScribbleTAS** [comment]
> this error usually means that your job doesn't have the checks: write permission.

This was exactly the issue here. I checked the readme, saw the section with the permissions and immediately put two and two together on why my workflow was failing, even for upload v3.

**Comment by AliakseiT** [comment]
For all who is affected by the [CVE-2024-42471](https://github.com/actions/toolkit/security/advisories/GHSA-6q32-hq47-5qq3), fix permissions as explained here https://stackoverflow.com/a/78679516/5608789

**Comment by JojOatXGME** [comment]
FYI, `dorny/test-reporter@v1` runs together with `actions/upload-artifact@v4` at [JetBrains/gradle-changelog-plugin](https://github.com/JetBrains/gradle-changelog-plugin/actions/workflows/test-report.yml). All the runs started to fail after `actions/upload-artifact` has been updated to version 4, but after #444 has been resolved, all the runs were green again.

In short: I still believe this issue is already resolved since version 1.9.1.

**Comment by HofmeisterAn** [comment]
Thanks, @JojOatXGME, for the info. After pinning the version to `1.9.1`, I was able to remove `actions/upload-artifact@v3` and update it to version `4` as well 👍.

**Comment by Gedochao** [comment]
https://github.com/VirtusLab/scala-cli/pull/2701
Seems to indeed be fixed, test reports get generated fine now.
Closing this.
